### PR TITLE
Allow fixtures YAML files to set the model class in the file itself

### LINKED ISSUE #9516: Rake db:fixtures:load - can't set fixture_class_name.
I have a many to many relation using a through table:

```ruby
class User < ActiveRecord::Base
  has_many :badges, :through => :users_badges 
end

class UsersBadges < ActiveRecord::Base
  belongs_to :user
  belongs_to :badge
  belongs_to :badgesource, :polymorphic => true
end

class Badge < ActiveRecord::Base
  attr_accessible :points, :description
  has_many :users, :through => :users_badges
end
```

And a fixtures file named `users_badges.yaml` inside of test/fixtures.

The first issue was really that it was recognizing the class name for my UsersBadges model as UsersBadge ( which I'm not sure is correct - I couldn't find documentation on this).

So - the workaround was to add `set_fixture_class` before the fixtures are loaded inside of test_helper.rb

Unfortunately, this only works in the test context. It's nice to interact with the fixtures in the console but running rake:db:fixtures:load does the same naming conversion translation. 

When it gets the class name wrong, it doesn't correctly grok the fixture file. My fixtures are related using labels. The SQL insert statement isn't adjusted to use ids and I get:

table `users_badges` has no column named user: 

```sql
INSERT INTO "users_badges" ("user", "badge", "badgesource") VALUES ('user1', 'mood_badge', 'rocking_out (Generic)').
```

The relevant code is around `lib/active_record/fixtures.rb:492`

### PR DESCRIPTION
**Pull request by repinel**
Currently, `set_fixture_class` is only available using the `TestFixtures` concern and it is ignored for `rake db:fixtures:load`. Using the correct model class, it is possible for the fixture load to also load the associations from the YAML files (e.g., `:belongs_to` and `:has_many`).

If the PR goes forward, should we consider deprecating `set_fixture_class`?

Fixes #9516.

### CODE REVIEW

**Comment by repinel** [comment]
@senny This would be the idea that we were discussing on #9516.

/cc @jeremy

**Comment by repinel** [comment]
@senny @jeremy Any thoughts on this?

**Comment by jeremy** [comment]
Using ERB within yaml to set metadata feels surprising.

How about using a special key to set fixture metadata?
```yaml
_fixture:
  model_class: User
david:
  name: David
```

**Comment by repinel** [comment]
@jeremy That is close to my first thought. IMO, it seems good.

Should I update the RP?

**Comment by senny** [comment]
@repinel yes, please do.

**Comment by repinel** [comment]
Closing and reopening to run tests.

File: activerecord/test/schema/schema.rb [hunk 1]
```diff
@@ -370,6 +374,11 @@ def except(adapter_names_to_exclude)
     t.integer :ideal_reference_id
   end
 
+  create_table :joke_comments, force: true do |t|
```
Comment (senny) [inline activerecord/test/schema/schema.rb]: is it necessary to create new tables to test this feature? Can't we reuse some tables that already exist? We try to not add more global models for single test scenarios like this.

**Comment by senny** [comment]
I think `set_fixture_class` should have priority over `_fixture:model_class`. This way you could configure the class inside the `.yml` file and still overwrite that inside a test case.

File: activerecord/lib/active_record/fixtures.rb [hunk 2]
```diff
@@ -395,6 +395,22 @@ class FixtureClassNotFound < ActiveRecord::ActiveRecordError #:nodoc:
...
   # Any fixture labeled "DEFAULTS" is safely ignored.
+  #
+  # == Support to custom model class
+  #
+  # You can also set the model class in your fixtures YAML file.
+  # This is helpful when fixtures are being loaded outside tests and
+  # you cannot use +set_fixture_class+, e.g., when running
+  # <tt>rake db:fixtures:load</tt>.
+  #
+  # To load the fixtures file `accounts.yml` as the `User` model, use:
```
Comment (senny) [inline activerecord/lib/active_record/fixtures.rb]: let's give the full path here: `test/fixtures/accounts.yml`.

File: activerecord/lib/active_record/fixtures.rb [hunk 3]
```diff
@@ -578,12 +594,16 @@ def initialize(connection, name, class_name, path, config = ActiveRecord::Base)
...
-        @model_class = class_name.safe_constantize if class_name
+      self.model_class = class_name
+
+      @fixtures = read_fixture_files(path)
+
+      # Sets the custom model class to all loaded fixtures.
+      fixture_settings = @fixtures.delete('_fixture')
+      if fixture_settings && (custom_model_class = fixture_settings['model_class'])
+        self.model_class = custom_model_class
+        @fixtures.values.each { |fixture| fixture.model_class = self.model_class }
```
Comment (senny) [inline activerecord/lib/active_record/fixtures.rb]: I don't like that we create `ActiveRecord::Fixture` objects with the "wrong" `model_class` and then re-assign here. Can we find a way that moves knowledge about `_fixture:model_class` into `Fixture::File` so that we can create `Fixture` objects with the right `model_class` from the get go?

**Comment by repinel** [comment]
@senny I updated the code. Please review, thanks!

File: activerecord/lib/active_record/fixtures.rb [hunk 4]
```diff
@@ -790,7 +814,7 @@ class FixtureError < StandardError #:nodoc:
     class FormatError < FixtureError #:nodoc:
     end
 
-    attr_reader :model_class, :fixture
+    attr_reader :fixture, :model_class
```
Comment (senny) [inline activerecord/lib/active_record/fixtures.rb]: why this change?

File: activerecord/lib/active_record/fixtures.rb [hunk 5]
```diff
@@ -761,13 +773,25 @@ def column_names
         @column_names ||= @connection.columns(@table_name).collect(&:name)
       end
 
-      def read_fixture_files(path, model_class)
+      def model_class=(class_name)
```
Comment (senny) [inline activerecord/lib/active_record/fixtures.rb]: Do we still need this extraction? It looks like we could revert that part back to how it's working on `master`?

File: activerecord/lib/active_record/fixtures.rb (hunk 4 above)
Comment (repinel) [inline activerecord/lib/active_record/fixtures.rb]: I missed this one. I'll revert the line.

File: activerecord/lib/active_record/fixtures.rb (hunk 5 above)
Comment (repinel) [inline activerecord/lib/active_record/fixtures.rb]: I was avoiding setting `model_class` in `read_fixture_files` directly to the instance variable and having to `safe_constantize` there. It's easy to revert if preferable.

File: activerecord/lib/active_record/fixture_set/file.rb [hunk 6]
```diff
@@ -18,12 +18,25 @@ def self.open(file)
...
       end
 
       def each(&block)
         rows.each(&block)
       end
 
+      def model_class
+        return @model_class if @model_class
+
+        rows.delete_if do |fixture_name, row|
```
Comment (senny) [inline activerecord/lib/active_record/fixture_set/file.rb]: This implementation is based on the fact that `model_class` will be called before `each`, otherwise you'd get a fixture for `_fixture`. We should probably change it to behave consistently regardless of call order. Meaning that we have to treat the `_fixture` key special that it won't ever be part of rows.

This could look something like this (pseudocode):

```ruby
def model_class
  raw_content["_fixture"]
  # ...
end

def rows
  raw_content.except "_fixture"
end

def raw_content
  # read yaml-file and cache contents
end
```

**Comment by repinel** [comment]
@senny I re-implemented the `AR::FixtureSet::File#model_class` method as suggested :smiley:

**Comment by senny** [comment]
:tada:

@repinel thanks for your work and patience :yellow_heart:. Finally found the time to finish this up. I made some minor changes directly in the merge commit (591a0bb87fff7583e01156696fbbf929d48d3e54)

**Comment by repinel** [comment]
@senny Looks awesome! Thanks for reviewing it :heart:
